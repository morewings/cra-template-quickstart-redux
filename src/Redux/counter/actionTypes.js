export const INCREMENT_COUNTER = 'INCREMENT_COUNTER';
