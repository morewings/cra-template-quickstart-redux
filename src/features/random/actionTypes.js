export const GET_RANDOM_NUMBER = 'GET_RANDOM_NUMBER';
