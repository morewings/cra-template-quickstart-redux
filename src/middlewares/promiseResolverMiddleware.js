/** Custom middleware function to resolve Promise given as an action payload property.
 * @see https://redux.js.org/understanding/history-and-design/middleware
 */
export const promiseResolverMiddleware = store => next => action => {
  if (typeof action?.payload?.then !== 'function') {
    /** Pass action to next middleware and exit, if payload is not a Promise */
    return next(action);
  }
  action.payload.then(
    response => {
      /** Use store.dispatch here and below to dispatch a new action through all
       * middlewares instead just pass it to next middleware */
      store.dispatch({
        type: `${action.type}_FULFILLED`,
        payload: response,
      });
    },
    () => {
      store.dispatch({
        type: `${action.type}_REJECTED`,
      });
    }
  );
  /** Pass *_PENDING action to next middleware */
  return next({type: `${action.type}_PENDING`});
};
