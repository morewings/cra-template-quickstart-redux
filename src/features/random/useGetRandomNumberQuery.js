import {useCallback} from 'react';
import {useDispatch} from 'react-redux';
import axios from 'axios';
import config from 'config';
import {GET_RANDOM_NUMBER} from './actionTypes';

const useGetRandomNumberQuery = () => {
  const dispatch = useDispatch();
  return useCallback(
    () =>
      dispatch({
        type: GET_RANDOM_NUMBER,
        payload: axios.get(config.randomAPI),
      }),
    [dispatch]
  );
};

export default useGetRandomNumberQuery;
