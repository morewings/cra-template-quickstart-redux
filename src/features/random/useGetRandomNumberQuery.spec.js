import React from 'react';
import {Provider} from 'react-redux';
import axios from 'axios';
import configureStore from 'redux-mock-store';
import {waitFor} from '@testing-library/react';
import {renderHook} from '@testing-library/react-hooks';
import {promiseResolverMiddleware} from '../../middlewares/promiseResolverMiddleware';
import {GET_RANDOM_NUMBER} from './actionTypes';
import useGetRandomNumberQuery from './useGetRandomNumberQuery';

jest.mock('axios');

describe('features > counter > useGetRandomNumberQuery', () => {
  /** Create mock store with middlewares */
  const mockStore = configureStore([promiseResolverMiddleware]);

  const store = mockStore({
    random: {
      isLoading: false,
      hasError: false,
      isFulfilled: false,
    },
  });

  it('returns function', () => {
    /**
     * Render hook, using testing-library utility
     * @see https://react-hooks-testing-library.com/reference/api#renderhook
     */
    const {result} = renderHook(() => useGetRandomNumberQuery(), {
      wrapper: ({children}) => <Provider store={store}>{children}</Provider>,
    });

    expect(result.current).toBeInstanceOf(Function);
  });

  describe('gets number', () => {
    afterEach(() => {
      store.clearActions();
    });

    /** Note that tests functions are async */
    it(`handles successful API query`, async () => {
      const {result} = renderHook(() => useGetRandomNumberQuery(), {
        wrapper: ({children}) => <Provider store={store}>{children}</Provider>,
      });

      /** Mock response from API */
      const response = 6;

      /**
       * Mock axios successful response
       * @see https://www.robinwieruch.de/axios-jest
       */
      axios.get.mockImplementationOnce(() => Promise.resolve({data: response}));

      /**
       * Wait until async action finishes
       */
      await result.current();

      /** First dispatched action should have _PENDING suffix */
      expect(store.getActions()[0]).toEqual({
        type: `${GET_RANDOM_NUMBER}_PENDING`,
      });

      await waitFor(() => {
        /** Second dispatched action should have _FULFILLED suffix */
        expect(store.getActions()[1].type).toEqual(
          `${GET_RANDOM_NUMBER}_FULFILLED`
        );
        /** Second dispatched action should deliver response from API */
        expect(store.getActions()[1].payload.data).toEqual(response);
      });
    });

    it(`handles rejected API query`, async () => {
      const {result} = renderHook(() => useGetRandomNumberQuery(), {
        wrapper: ({children}) => <Provider store={store}>{children}</Provider>,
      });

      /**
       * Mock axios rejected response
       * @see https://www.robinwieruch.de/axios-jest
       */
      axios.get.mockImplementationOnce(() => Promise.reject(new Error('')));

      /**
       * Wait until async action finishes
       */
      await result.current();

      /** First dispatched action should have _PENDING suffix */
      expect(store.getActions()[0]).toEqual({
        type: `${GET_RANDOM_NUMBER}_PENDING`,
      });

      await waitFor(() => {
        /** Second dispatched action should have _REJECTED suffix */
        expect(store.getActions()[1].type).toEqual(
          `${GET_RANDOM_NUMBER}_REJECTED`
        );
      });
    });
  });
});
