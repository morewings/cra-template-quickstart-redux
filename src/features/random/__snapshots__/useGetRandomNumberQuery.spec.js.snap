// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 1`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 2`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 3`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 4`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 5`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 6`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 7`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 8`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 9`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 10`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 11`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 12`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 13`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 14`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 15`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 16`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 17`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 18`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 19`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 20`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 21`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 22`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 23`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 24`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 25`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 26`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 27`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 28`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 29`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 30`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 31`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 32`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 33`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 34`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 35`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 36`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 37`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 38`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 39`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 40`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 41`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 42`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 43`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 44`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 45`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 46`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 47`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 48`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 49`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 50`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 51`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 52`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 53`] = `undefined`;

exports[`features > counter > useGetRandomNumberQuery gets number it handles API fetching errors 54`] = `undefined`;
